types/tablet_v2/wlr_tablet_v2.c

```diff
@@ -0,0 +1,309 @@
+#ifndef _POSIX_C_SOURCE
+#define _POSIX_C_SOURCE 200809L
+#endif
+
+#include <assert.h>
+#include <libinput.h>
+#include <stdlib.h>
+#include <string.h>
+#include <wayland-server.h>
+#include <types/wlr_tablet_v2.h>
+#include <wlr/config.h>
+#include <wlr/types/wlr_seat.h>
+#include <wlr/types/wlr_tablet_tool.h>
+#include <wlr/types/wlr_tablet_v2.h>
+#include <wlr/util/log.h>
+#include "tablet-unstable-v2-protocol.h"
+
+struct wlr_tablet_manager_client_v2 {
+	struct wl_list link;
+	struct wl_client *client;
+	struct wl_resource *resource;
+	struct wlr_tablet_manager_v2 *manager;
+
+	struct wl_listener client_destroy;
+	struct wl_list tablet_seats; // wlr_tablet_seat_client_v2::link
+};
+
+static void handle_wlr_seat_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_tablet_seat_v2 *seat =
+		wl_container_of(listener, seat, seat_destroy);
+
+	wl_list_remove(&seat->link);
+	wl_list_remove(&seat->seat_destroy.link);
+
+	struct wlr_tablet_seat_client_v2 *client;
+	struct wlr_tablet_seat_client_v2 *tmp;
+	wl_list_for_each_safe(client, tmp, &seat->clients, seat_link) {
+		wlr_tablet_seat_client_v2_destroy(client->resource);
+	}
+}
+
+static struct wlr_tablet_seat_v2 *create_tablet_seat(
+		struct wlr_tablet_manager_v2 *manager,
+		struct wlr_seat *wlr_seat) {
+	struct wlr_tablet_seat_v2 *tablet_seat =
+		calloc(1, sizeof(struct wlr_tablet_seat_v2));
+	if (!tablet_seat) {
+		return NULL;
+	}
+
+	tablet_seat->manager = manager;
+	tablet_seat->wlr_seat = wlr_seat;
+
+	wl_list_init(&tablet_seat->clients);
+
+	wl_list_init(&tablet_seat->tablets);
+	wl_list_init(&tablet_seat->tools);
+	wl_list_init(&tablet_seat->pads);
+
+	tablet_seat->seat_destroy.notify = handle_wlr_seat_destroy;
+	wl_signal_add(&wlr_seat->events.destroy, &tablet_seat->seat_destroy);
+
+	wl_list_insert(&manager->seats, &tablet_seat->link);
+	return tablet_seat;
+}
+
+struct wlr_tablet_seat_v2 *get_or_create_tablet_seat(
```

Comment:
Should be `static`

Reply:
It's used by the other files I split the implementation into, but shouldn't be API.
What's the proper naming scheme for them? Afaik it shouldn't have the `wlr_` prefix.

Reply:
This *is* in the private header :) `include/types/` not `/include/wlr/tablet/` :+1:

Reply:
Ah wait, no. I got confused for a second. And I still didn't get any answer about what the proper non-colliding naming scheme would be here

Reply:
`tablet_seat_get_or_create` seems reasonable